### PR TITLE
Improve raft write performance by utilizing FSM Batching

### PR DESCRIPTION
**Pull request by briankassouf**
This PR depends on hashicorp/raft#364 and hashicorp/go-raftchunking#2. Those PRs enable FSM batching capabilities. This PR implements the new interface that is necessary to provide batching support. Once those PRs are merged the `go.mod` file here should be updated.

This PR improves multi-writer performance for the raft storage backend. Prior to this change the underlying raft library would serially apply logs to the FSM. When the FSM is all in-memory this isn't an issue. However since Vault's FSM persists data to a boltdb file we were seeing head of line blocking happening for every update. Now the FSM will batch updates to the same degree as log batching (defaults to 64 items per batch). 

Cursory benchmark data before/after this change (100 threads, 100 connections):
```
Before: 41.45 req/sec
After: 547.59 req/sec
```

### CODE REVIEW

File: physical/raft/fsm.go [hunk 1]
```diff
@@ -404,57 +384,99 @@ func (f *FSM) Transaction(ctx context.Context, txns []*physical.TxnEntry) error
...
+	var latestConfiguration *ConfigurationValue
+	commands := make([]interface{}, 0, len(logs))
+	for _, log := range logs {
+		switch log.Type {
+		case raft.LogCommand:
+			command := &LogData{}
+			err := proto.Unmarshal(log.Data, command)
+			if err != nil {
+				f.logger.Error("error proto unmarshaling log data", "error", err)
+				panic(len(log.Data))
```
Comment (michelvocks) [inline physical/raft/fsm.go]: Is the `panic` here intended? Will the panic below executed at all?

File: physical/raft/fsm.go [hunk 2]
```diff
@@ -472,13 +494,32 @@ func (f *FSM) Apply(log *raft.Log) interface{} {
...
+	}
+
+	// If one or more configuration changes were processed, store the latest one.
+	if latestConfiguration != nil {
+		f.latestConfig.Store(latestConfiguration)
 	}
 
-	return &FSMApplyResponse{
-		Success: true,
+	// Build the responses. The logs array is used here to esnure we reply to
```
Comment (michelvocks) [inline physical/raft/fsm.go]:
```suggestion
	// Build the responses. The logs array is used here to ensure we reply to
```

File: physical/raft/fsm.go (hunk 1 above)
Comment (briankassouf) [reply physical/raft/fsm.go]: Sure isn't, left in from some debugging

Comment (briankassouf) [inline physical/raft/fsm.go]:
```suggestion
```

**Pull request merged**